src/main/java/schemadefiner/retrieveschema/JdbcCrawler.java

```diff
@@ -0,0 +1,75 @@
+import schemacrawler.schema.Schema;
+import schemacrawler.schema.Table;
+import schemacrawler.schemacrawler.SchemaCrawlerException;
+import schemacrawler.schemacrawler.SchemaCrawlerOptions;
+import schemacrawler.tools.utility.SchemaCrawlerUtility;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class JdbcCrawler implements SchemaDefiner {
+
+    private static final Logger log = LoggerFactory.getLogger(JdbcCrawler.class);
+
+    private String user;
+    private String password;
+    private String jdbcUrl;
+    private final Supplier<Connection> connectionSupplier;
+    private final SchemaCrawlerOptions schemaCrawlerOptions;
+    private String schemaName;
+    private List<String> tableTypes;
+
+    public JdbcCrawler(String jdbcUrl, Supplier<Connection> connectionSupplier, SchemaCrawlerOptions schemaCrawlerOptions, String schemaName, List<String> tableTypes) {
+        this.jdbcUrl = jdbcUrl;
+        this.connectionSupplier = connectionSupplier;
+        this.schemaCrawlerOptions = schemaCrawlerOptions;
+        this.schemaName = schemaName;
+        this.tableTypes = tableTypes;
+    }
+
+    @Override
+    public StreamlinerCatalog retrieveSchema() throws SQLException, InterruptedException {
+        if (jdbcUrl.startsWith("jdbc:oracle")) {
+            log.info("ByPassing Oracle from JdbcCrawler");
+            StopWatch retrieveTablesTimer = StopWatch.createStarted();
+            // there are few tables used in getOracleCatalog() that are not public.
+            StreamlinerCatalog catalog = StreamlinerSchemaCrawler.getOracleCatalog(new StreamlinerSchemaCrawler.QueryHandler(connectionSupplier), schemaName, tableTypes);
+            retrieveTablesTimer.stop();
+            log.info(String.format("Retrieve Tables took %s", retrieveTablesTimer.formatTime()));
+            return catalog;
+        }
+        try (Connection connection = connectionSupplier.get()) {
+            Map<Schema, List<Table>> schemaTableMap = new HashMap<>();
+            Catalog catalog = SchemaCrawlerUtility.getCatalog(connection, schemaCrawlerOptions);
+            for (Schema schema : catalog.getSchemas()) {
+                List<Table> tables = schemaTableMap.get(schema);
+                if (tables == null) {
+                    tables = new ArrayList<>();
+                    schemaTableMap.put(schema, tables);
+                }
+                tables.addAll(catalog.getTables(schema));
+            }
+            return new StreamlinerCatalog(catalog.getJdbcDriverInfo().getDriverClassName(),
+                    new ArrayList<>(catalog.getSchemas()), schemaTableMap);
+        } catch (SchemaCrawlerException | SQLException e){
+            log.error("Error in JdbcCrawler: {}", e.getMessage());
+            throw new RuntimeException("Error in JdbcCrawler: {}",e);
+        }
+
```

Comment:
Can we call getCatalog() of StreamlinerSchemaCrawler directly instead of copying the code because getOracleCatalog() methods we can not copy since few schemacrawler tables can not be accessed through other packages .

Reply:
I don't quite follow, but ideally we aren't copying code. You are free to move code around as you please. If you need to place something in the schemacrawler package or make something public over there, that is ok.

We can discuss after tomorrow's demo if you have further questions.